### PR TITLE
#311 mailet

### PR DESCRIPTION
**Pull request by tieugiang91**
Implement mailet to index recipients when email sent.

### CODE REVIEW

**Comment by chibenwa** [comment]
We do not need ES implementation to write the mailet.

Can you please drop that commit from here?

File: tmail-backend/jmap/extensions-es/src/test/java/com/linagora.tmail.james.jmap.model/ESEmailAddressContactSearchTest.java [hunk 1]
```diff
@@ -1,12 +1,11 @@
 package com.linagora.tmail.james.jmap.model;
 
 import com.linagora.tmail.james.jmap.ESEmailAddressContactSearchEngine;
-import com.linagora.tmail.james.jmap.model.EmailAddressContactSearchEngine;
 import org.apache.james.backends.es.v7.DockerElasticSearchExtension;
 import org.apache.james.backends.es.v7.ReactorElasticSearchClient;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
-public class ESEmailAdressContactSearchTest implements EmailAddressContactSearchEngineContract{
+public class ESEmailAddressContactSearchTest implements EmailAddressContactSearchEngineContract{
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions-es/src/test/java/com/linagora.tmail.james.jmap.model/ESEmailAddressContactSearchTest.java:8]: Unrelated changr

File: tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.apache.mailet.base.GenericMailet;
+import reactor.core.publisher.Flux;
+
+import javax.mail.MessagingException;
+import java.util.UUID;
+
+public class TmailEmailContactSearchEngineMailet extends GenericMailet {
+    final EmailAddressContactSearchEngine emailAddressContactSearchEngine;
+
+    public TmailEmailContactSearchEngineMailet(EmailAddressContactSearchEngine emailAddressContactSearchEngine) {
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java:16]: Missing @inject annotation

File: tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Flux.fromIterable(mail.getRecipients())
+            .map(recipient -> emailAddressContactSearchEngine
+                 .index(AccountId.fromString(mail.getMaybeSender().asString()),
+                        new EmailAddressContact(UUID.fromString(recipient.getLocalPart()),
+                                                recipient.getLocalPart()+"@"+recipient.getDomain())));
+    }
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java:27]: Li would like we rely o  strong typing here...

**Review by chibenwa — COMMENTED** [review]
Create mailet is not a good enough commit titlr

File: tmail-backend/jmap/extensions-es/src/test/java/com/linagora.tmail.james.jmap.model/ESEmailAddressContactSearchTest.java (hunk 1 above)
Comment (tieugiang91) [reply tmail-backend/jmap/extensions-es/src/test/java/com/linagora.tmail.james.jmap.model/ESEmailAddressContactSearchTest.java:8]: May I just put it here, as suddenly found out that dictation error so couldn't help editing :(

Comment (Arsnael) [reply tmail-backend/jmap/extensions-es/src/test/java/com/linagora.tmail.james.jmap.model/ESEmailAddressContactSearchTest.java:8]: You should change it in the related PR or separated one, or separated commit. This has nothing to do with the mailet :)

File: tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        this.emailAddressContactSearchEngine = emailAddressContactSearchEngine;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Flux.fromIterable(mail.getRecipients())
+            .map(recipient -> emailAddressContactSearchEngine
+                 .index(AccountId.fromString(mail.getMaybeSender().asString()),
+                        new EmailAddressContact(UUID.fromString(recipient.getLocalPart()),
+                                                recipient.getLocalPart()+"@"+recipient.getDomain().asString())));
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java:28]: Can we avoid string manipulation outside of the domain model?

This breaks the encapsulation brought in by object origented programming.

File: tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    public TmailEmailContactSearchEngineMailet(EmailAddressContactSearchEngine emailAddressContactSearchEngine) {
+        this.emailAddressContactSearchEngine = emailAddressContactSearchEngine;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        Flux.fromIterable(mail.getRecipients())
+            .map(recipient -> emailAddressContactSearchEngine
+                 .index(AccountId.fromString(mail.getMaybeSender().asString()),
+                        new EmailAddressContact(UUID.fromString(recipient.getLocalPart()),
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/java/com/linagora/tmail/james/jmap/TmailEmailContactSearchEngineMailet.java:27]:
```
UUID.fromString(recipient.getLocalPart())
```

bob@attacker.com would then have the same UUID than bob@victim.com, I could use this to override auto-completion proposals.

Also, as a best practice **we should NEVER let the caller of an API choose an id** as this is the door open for overrides, inconsistencies and all the sort of unpleasant bugs.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala [hunk 6]
```diff
@@ -10,6 +11,9 @@ import scala.jdk.CollectionConverters.CollectionHasAsScala
 
 case class EmailAddressContact (id: UUID, address: String) {
   def contains(part: String): Boolean = address.contains(part)
+  def this(mailAddress: MailAddress){
+    this(UUID.fromString(mailAddress.hashCode().toString), mailAddress.asString())
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/EmailAddressContact.scala:15]:
```
    UUID.fromString(mailAddress.hashCode().toString)
```

This likely do not have the expected entropy...

**Comment by chibenwa** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (default) on project jmap-extensions: You have 7 Checkstyle violations. -> [Help 1]
```

Not compiling....

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

**Comment by vttranlina** [comment]
Resolve at https://github.com/linagora/tmail-backend/pull/387

**Pull request closed without merging**